src/libstd/collections/hash/raw/mod.rs

```diff
@@ -0,0 +1,1212 @@
+    }
+
+    /// Searches for an empty or deleted bucket which is suitable for inserting
+    /// a new element.
+    ///
+    /// There must be at least 1 empty bucket in the table.
+    #[inline]
+    fn find_insert_slot(&self, hash: u64) -> usize {
+        for pos in self.probe_seq(hash) {
+            unsafe {
+                let group = Group::load(self.ctrl(pos));
+                if let Some(bit) = group.match_empty_or_deleted().lowest_set_bit() {
+                    let result = (pos + bit) & self.bucket_mask;
+
+                    // In tables smaller than the group width, trailing control
+                    // bytes outside the range of the table are filled with
+                    // DELETED entries. These will unfortunately trigger a
+                    // match, but once masked will point to a full bucket that
+                    // is already occupied. We detect this situation here and
+                    // perform a second scan starting at the begining of the
+                    // table. This second scan is guaranteed to find an empty
+                    // slot (due to the load factor) before hitting the trailing
+                    // control bytes (containing DELETED).
+                    if unlikely(is_full(*self.ctrl(result))) {
+                        debug_assert!(self.bucket_mask < Group::WIDTH);
+                        debug_assert_ne!(pos, 0);
+                        return Group::load_aligned(self.ctrl(0))
+                            .match_empty_or_deleted()
+                            .lowest_set_bit_nonzero();
+                    } else {
+                        return result;
+                    }
+                }
+            }
+        }
+
+        // probe_seq never returns.
+        unreachable!();
+    }
+
+    /// Marks all table buckets as empty without dropping their contents.
+    #[inline]
+    pub fn clear_no_drop(&mut self) {
+        if self.bucket_mask != 0 {
+            unsafe {
+                self.ctrl(0)
+                    .write_bytes(EMPTY, self.buckets() + Group::WIDTH);
+            }
+        }
+        self.items = 0;
+        self.growth_left = bucket_mask_to_capacity(self.bucket_mask);
+    }
+
+    /// Removes all elements from the table without freeing the backing memory.
+    #[inline]
+    pub fn clear(&mut self) {
+        // Ensure that the table is reset even if one of the drops panic
+        let self_ = guard(self, |self_| self_.clear_no_drop());
+
+        if mem::needs_drop::<T>() {
+            unsafe {
+                for item in self_.iter() {
+                    item.drop();
+                }
+            }
+        }
+    }
+
+    /// Shrinks the table to fit `max(self.len(), min_size)` elements.
+    #[inline]
+    pub fn shrink_to(&mut self, min_size: usize, hasher: impl Fn(&T) -> u64) {
+        let min_size = usize::max(self.items, min_size);
+        if self.bucket_mask != 0 && bucket_mask_to_capacity(self.bucket_mask) >= min_size * 2 {
```

Comment:
We should add a `debug_assert!` here for clarity checking the inputs. We do compile the standard library with debug assertions enabled in the CI tests. Users can't easily do this yet, but maybe they will be able to do so one day.

Reply:
There's no benefit in having a debug assert since this is an overflow (which already debug asserts).

Reply:
> There's no benefit in having a debug assert since this is an overflow (which already debug asserts).

To me there is a huge benefit since I find:

```rust
debug_assert!(
    min_size <= usize::max_value() / 2, 
    "user is responsible of making sure that min_size does not overflow"
);
```

infinitely more clear than just using `min_size * 2` and hoping that the reader realizes what's going on.